design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,381 @@
+
+This all means, the following function may have to change:
+
+```python
+def set_virtual_ref(
+    self, key: str, location: str, offset: int, length: int
+) -> None: ...
+```
+
+### Credentials
+
+Credentials are a read time concern. When the user writes the virtual chunk, they don't necessarily know what credentials set will be used to read it.
+
+Because of requirements 1 and 2, when users open a repo they need to know what sets of credentials they have to provide to be able to access all the chunks. This is not immediately obvious, Icechunk should provide a way to introspect a repo and inform the user what credentials they need. Then the user can setup the Repo instance appropriately with the credentials.
+
+### Differences between object stores
+
+Many object stores support a subset of the S3 wire protocol to read and write objects. This seems to simplify the task, because Icechunk could talk to all those platforms with the same "client code". But there are many details that make this theoretical approach not work very well in practice.
+
+* The level of conformity with the protocol varies widely. Icechunk doesn't use very advanced features of the S3 protocol but this could change in the future. Even today, Icechunk already uses, for example, conditional requests (both writes and reads).
+* The ideal client settings for different platforms is different. Some object stores allow more aggressive retries than others, or more concurrency. In general, performance optimization needs to be targeted to a specific platform sometimes.
+* Encryption and checksumming capabilities changes widely too between platforms.
+* Error responses are not uniform, some tweaking is needed sometimes to extract the most relevant information.
+* Authentication protocols, and credential refresh varies widely.
+
+The more Icechunk knows about the specifics of the object store platform the better job it can do at fetching chunks. Since all the information for the virtual chunks is stored in the dataset at write time, it's important we store enough, to future proof for changes and improvements to the different platforms.
+
+All that said, good defaults and usability are very important. The ideal situation is: virtual chunks that work by default, and can be made as efficient as possible with a small amount of work.
+
+## Design
+
+### Checksum
+
+In order to support requirement 4, we add an optional `checksum` argument to `set_virtual_ref`.
```

Comment:
```suggestion
In order to support requirement 4, we add an optional `verification` argument to `set_virtual_ref`.
```

I'd prefer to not call this `checksum` since in reality it is more general than that.